### ISSUE TITLE
Maher's solutions

### ISSUE THREAD

**Issue by MaherAssaf19** [MIT-Emerging-Talent/ET6-foundations-group-17#22]
This Python script analyzes housing prices using a predefined dataset and performs key tasks such as data preprocessing, exploratory data analysis, predictive modeling, and visualization. Here's a breakdown of its functionality:

1. Data Loading
The script contains an embedded dataset of housing attributes such as:

SquareFeet: Total square footage of the house.
Bedrooms: Number of bedrooms.
Bathrooms: Number of bathrooms.
YearBuilt: Year the house was constructed.
LocationScore: A score representing the desirability of the house's location.
Price: The sale price of the house.
The dataset is loaded into a Pandas DataFrame for further processing.

2. Data Preprocessing
The preprocess_data function cleans the dataset by removing any missing values. This ensures the data is consistent and ready for analysis.

3. Exploratory Data Analysis
The analyze_data function provides:

Summary Statistics: Basic metrics (mean, standard deviation, etc.) for numerical features.
Visualizations: Pair plots to display relationships among variables and highlight patterns or trends.
4. Model Training
The train_model function uses:

Features: SquareFeet, Bedrooms, Bathrooms, YearBuilt, LocationScore.
Target: Price.
A linear regression model is trained using a split of the data into training and testing subsets (80/20 split). The trained model can predict housing prices based on the input features.

5. Model Evaluation
The evaluate_model function assesses the model's performance using:

Mean Squared Error (MSE): Measures prediction accuracy.
R-squared (R²): Indicates the proportion of variance in the target variable explained by the features.
6. Visualization
The visualize_results function provides:

Actual vs Predicted Prices: A scatter plot showing how well the model's predictions match the actual prices.
Feature Coefficients: A bar chart highlighting the importance of each feature in the regression model.
Key Features of the Script
Modular Design: Each function handles a specific part of the workflow, making the code reusable and easy to maintain.
Embedded Dataset: The script includes a hardcoded dataset for immediate execution.
Visualization: Provides insights into model performance and feature significance.
This script is ready to use, offering a complete framework for housing price analysis. Let me know if you need additional modifications or detailed explanations!

Here is the code to the solution>👇

# housing_prices_analysis_solution.py

def load_data():
    """
    Load a predefined housing dataset.

    Returns:
        pd.DataFrame: Loaded dataset as a Pandas DataFrame.
    """
    import pandas as pd
    from io import StringIO

    # Embedded dataset
    data = """SquareFeet,Bedrooms,Bathrooms,YearBuilt,LocationScore,Price
1500,3,2,2000,85,300000
2000,4,3,2010,90,450000
1800,3,2,2005,88,350000
2400,4,3,2020,92,500000
1600,3,2,1995,80,280000
1200,2,1,1980,70,200000
"""
    return pd.read_csv(StringIO(data))

def preprocess_data(data):
    """
    Preprocess the housing dataset by handling missing values and extracting necessary features.
    """
    return data.dropna()

def analyze_data(data):
    """
    Perform exploratory data analysis on the dataset.
    """
    import matplotlib.pyplot as plt
    import seaborn as sns

    print("Dataset Summary:")
    print(data.describe())

    sns.pairplot(data[['SquareFeet', 'Bedrooms', 'Bathrooms', 'YearBuilt', 'LocationScore', 'Price']])
    plt.show()

def train_model(data):
    """
    Train a predictive model using the dataset.
    """
    from sklearn.model_selection import train_test_split
    from sklearn.linear_model import LinearRegression

    features = ['SquareFeet', 'Bedrooms', 'Bathrooms', 'YearBuilt', 'LocationScore']
    target = 'Price'

    X = data[features]
    y = data[target]

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    model = LinearRegression()
    model.fit(X_train, y_train)

    return model, X_test, y_test

def evaluate_model(model, X_test, y_test):
    """
    Evaluate the trained model using Mean Squared Error and R-squared metrics.
    """
    from sklearn.metrics import mean_squared_error, r2_score

    y_pred = model.predict(X_test)

    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)

    print(f"Mean Squared Error: {mse}")
    print(f"R-squared: {r2}")

def visualize_results(model, X_test, y_test):
    """
    Visualize the actual vs predicted prices and feature importance.
    """
    import matplotlib.pyplot as plt
    import pandas as pd

    # Actual vs Predicted Prices
    y_pred = model.predict(X_test)
    plt.figure(figsize=(10, 6))
    plt.scatter(y_test, y_pred, alpha=0.6, color='blue')
    plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], '--r', linewidth=2)
    plt.xlabel('Actual Prices')
    plt.ylabel('Predicted Prices')
    plt.title('Actual vs Predicted Prices')
    plt.show()

    # Feature Coefficients
    coefficients = pd.Series(model.coef_, index=X_test.columns)
    plt.figure(figsize=(8, 4))
    coefficients.plot(kind='bar', color='skyblue')
    plt.title('Feature Coefficients')
    plt.ylabel('Coefficient Value')
    plt.show()

if _name_ == "_main_":
    data = load_data()
    data = preprocess_data(data)
    analyze_data(data)
    model, X_test, y_test = train_model(data)
    evaluate_model(model, X_test, y_test)
    visualize_results(model, X_test, y_test)

**Comment by MaRia19280** [comment]
Resolved#22
